File: Algorithm_LevelOne/Remove_Nth_Node_From_End_of_List/Remove_Nth_Node_From_End_of_List.py
# 19. Remove Nth Node From End of List Python

# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution(object):
    def removeNthFromEnd(self, head, n):
        """
        :type head: ListNode
        :type n: int
        :rtype: ListNode
        """
        count = 1
        test =  head
        output = pre = ListNode(0, head)
        while test:
            if test.next is None:
                break
            test = test.next
            if count < n:
                count += 1
            else:
                pre = pre.next
        pre.next = pre.next.next if pre.next is not None else None
        return output.next

File: Algorithm_LevelOne/Remove_Nth_Node_From_End_of_List/Remove_Nth_Node_From_End_of_List.cc
// 19. Remove Nth Node From End of List C++

// Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode* output = new ListNode(0, head);
        ListNode* pre = output;
        while (head) {
            if (n > 0) {
                n--;
            } else {
                pre = pre->next;
            }
            head = head->next;
        }
        if (pre->next) pre->next = pre->next->next;
        pre = output->next;
        delete output;
        return pre;
    }
};
